doc/design/tensor_design.md

```diff
@@ -0,0 +1,173 @@
+  // No assigning!
+  Allocation &operator=(const Allocation &) = delete;
+
+  void *ptr() const;
+  void *end() const;
+  Device device() const;
+  size_t size() const;
+
+ private:
+  bool owned_;
+  void *ptr_;
+  size_t size_;
+  Device device_;
+};
+```
+
+`ptr_` points to the head of the memory piece and `size_` shows its length. `owned_` marks whether the memory piece is allocated by `allocation` itself, if so the memory will be freed when `allocation` is destructed.
+
+`Device` is something like Majel's `Place`. However, `Place` in Majel is an alias of `boost::variant`, while `Device` here is a certain class (can be specialized to `CpuDevice` or `GpuDevice`). `CpuDevice` and `GpuDevice` are exactly Majel's `CpuPlace` and `GpuPlace`, we rename them to fit the overall naming style.
+
+```cpp
+struct CpuDevice {
+  inline bool operator==(const CpuDevice &) const { return true; }
+  inline bool operator!=(const CpuDevice &) const { return false; }
+};
+
+struct GpuDevice {
+  GpuDevice(int d) : device_id(d) {}
+
+  inline bool operator==(const GpuDevice &o) const {
+    return device_id == o.device_id;
+  }
+  inline bool operator!=(const GpuDevice &o) const { return !(*this == o); }
+  GpuDevice() : GpuDevice(0) {}
+  int device_id;
+};
+```
+
+## Tensor
+
+`Tensor` is the combination of Majel's `Buffer` and `Array`.
+
+```cpp
+template <typename Device, typename T, int rank>
+class Tensor {
+ public:
+  // tensor with zero size and no memory
+  Tensor();
+  // allocates new densely packed tensor
+  Tensor(const Dim<rank> size, Device device);
+
+  // make a new tensor by another existing tensor
+  // new tensor and source tensor have the same numel but deferent rank
+  template <int src_rank>
+  Tensor(const Dim<rank> &size, Tensor<Device, T, src_rank> &src);
+
+  // '=' are not allowed, because users may be confused about
+  //     whether it's deep copy or shallow copy.
+  Tensor &operator=(const Tensor &src) = delete;
+
+  // return raw pointer to the data.
+  T *raw_ptr() const;
+
+  // return tensor size
+  Dim<rank> size() const;
+
+  // return the number of tensor elements
+  int numel() const;
+
+  // return tensor stride
+  Dim<rank> stride() const;
+
+  // return raw pointer to the 'idx'th element
+  T *index(const Dim<rank> &idx) const;
+
+  // resize tensor, data may be erased
+  void resize(const Dim<rank> &size);
+
+  // reshape tensor, data will be retained
+  void reshape(const Dim<rank> &size);
+
+ private:
+  std::shared_ptr<Allocation<Device>> allocation_;
+  Dim<rank> size_;
+  Dim<rank> stride_;
+  T *ptr_;
+};
+```
+
+The member variable `allocation_` points to the `Allocation` object where data are stored. However, one `Allocation` object can be shared by several tensors, so we need another raw pointer `ptr_` to indicate where is the head of **this** tensor's data. 
+
+`size_` and `stride_` are `Dim` object. Inspired from Majel, `Dim` is a struct template for indicating tensor size and element index:
+
+```cpp
```

Comment:
Here I only show a few parts of `Dim`.
You can check full definition of `Dim` in https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/dim.h